### PR TITLE
Fix syntax to set default target in aws-pcluster pipelines

### PR DESCRIPTION
**Pull request by stephenmsachs**

### CODE REVIEW

**Comment by stephenmsachs** [comment]
Fixes error like these (https://gitlab.spack.io/spack/spack/-/jobs/7270468) as the target is not spcified by the generate step machine but excplicilty set as `SPACK_TARGET_ARCH`.

**Comment by stephenmsachs** [comment]
Stuff failed in pipelines not affected by this PR. @spackbot re-run pipeline.

**Comment by stephenmsachs** [comment]
Stuff failed in pipelines not affected by this PR. @spackbot re-run pipeline.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 1]
```diff
@@ -736,7 +736,7 @@ deprecated-ci-build:
       - cd "${CI_PROJECT_DIR}" && curl -sOL https://raw.githubusercontent.com/spack/spack-configs/main/AWS/parallelcluster/postinstall.sh
       - sed -i -e "s/spack arch -t/echo ${SPACK_TARGET_ARCH}/g" postinstall.sh
       - /bin/bash postinstall.sh -fg
-      - spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
+      - spack config --scope site add "packages:all:target:\"${SPACK_TARGET_ARCH}\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Is this working? Package preferences expect an array of strings for targets, and this is a single target. Trying the same from the command line I get an error:
```console
$ spack config add packages:all:target:x86_64
==> Error: <unicode string>:1: 'x86_64' is not of type 'array'
```

**Review by alalazo — CHANGES_REQUESTED** [review]
Is there any reason we are using package preferences, instead of requirements (which have stronger semantic)?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml (hunk 1 above)
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Yes, it works for me. Maybe it's the escaped quotes? 
DSo you are saying the correct syntax would be `spack config --scope site add "packages:all:target:[\
"${SPACK_TARGET_ARCH}\"]"`?

Comment (alalazo) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Can you post the resulting YAML file?

Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]:
```
bash-4.2# spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
bash-4.2# cat etc/spack/packages.yaml
packages:
  [...]
  all:
    compiler: [nvhpc, gcc, clang]
    providers:
      blas: [armpl-gcc, openblas]
      fftw-api: [armpl-gcc, fftw]
      lapack: [armpl-gcc, openblas]
      mpi: [openmpi, mpich]
      scalapack: [netlib-scalapack]
    permissions:
      read: world
      write: user
    target:
    - target=neoverse_n1
```

Comment (alalazo) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ok, I'd like to get to the bottom of this:
Screenshot from 2023-06-08 15-17-25

I am using:

* **Spack:** 0.21.0.dev0 (315873cbd329a05f8cb944bcfb8c9ec2866493ff)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

Comment (alalazo) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Confirmed that in:
```console
docker run -it --rm "ghcr.io/spack/pcluster-amazonlinux-2:v2023-05-25"
```
it works as reported by @stephenmsachs :thinking: Need to check if it is the Python version being used (3.7 vs. 3.8). In any case:
```console
$ spack config --scope site add packages:all:target:"[${SPACK_TARGET_ARCH}]"
```
works on both platforms.

**Comment by stephenmsachs** [comment]
> Is there any reason we are using package preferences, instead of requirements (which have stronger semantic)?

Yes. Because we are building the compiler(s) for x86_64. AFAIK that contradicts a requirement for "all".

**Review by alalazo — APPROVED** [review]

**Review by alalazo — CHANGES_REQUESTED** [review]
It seems there's a conflict with `.gitlab-ci.yml` to be solved.

**Review by alalazo — APPROVED** [review]

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 2]
```diff
@@ -744,7 +744,7 @@ deprecated-ci-build:
       - sed -i.bkp s/"spack install gcc"/"spack install --cache-only --reuse gcc"/ postinstall.sh
       - diff postinstall.sh postinstall.sh.bkp || echo Done
       - /bin/bash postinstall.sh -fg
-      - spack config --scope site add "packages:all:target:\"${SPACK_TARGET_ARCH}\""
+      - spack config --scope site add "packages:all:target:\"[${SPACK_TARGET_ARCH}]\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]:
```suggestion
      - spack config --scope site add "packages:all:target:[${SPACK_TARGET_ARCH}]"
```

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml [hunk 3]
```diff
@@ -48,7 +48,7 @@ spack:
           - spack gpg trust /bootstrap/public-key
           - diff /bootstrap/postinstall.sh /bootstrap/postinstall.sh.bkp || echo Done
         - - /bin/bash "${SPACK_ARTIFACTS_ROOT}/postinstall.sh" -fg
-          - spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
+          - spack config --scope site add "packages:all:target:\"[${SPACK_TARGET_ARCH}]\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml]:
```suggestion
          - spack config --scope site add "packages:all:target:[${SPACK_TARGET_ARCH}]"
```

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml [hunk 4]
```diff
@@ -50,7 +50,7 @@ spack:
           - spack gpg trust /bootstrap/public-key
           - diff /bootstrap/postinstall.sh /bootstrap/postinstall.sh.bkp || echo Done
         - - /bin/bash "${SPACK_ARTIFACTS_ROOT}/postinstall.sh" -fg
-          - spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
+          - spack config --scope site add "packages:all:target:\"[${SPACK_TARGET_ARCH}]\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml]:
```suggestion
          - spack config --scope site add "packages:all:target:[${SPACK_TARGET_ARCH}]"
```

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_v1/spack.yaml [hunk 5]
```diff
@@ -50,7 +50,7 @@ spack:
           - spack gpg trust /bootstrap/public-key
           - diff /bootstrap/postinstall.sh /bootstrap/postinstall.sh.bkp || echo Done
         - - /bin/bash "${SPACK_ARTIFACTS_ROOT}/postinstall.sh" -fg
-          - spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
+          - spack config --scope site add "packages:all:target:\"[${SPACK_TARGET_ARCH}]\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_v1/spack.yaml]:
```suggestion
          - spack config --scope site add "packages:all:target:[${SPACK_TARGET_ARCH}]"
```

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-skylake/spack.yaml [hunk 6]
```diff
@@ -48,7 +48,7 @@ spack:
           - spack gpg trust /bootstrap/public-key
           - diff /bootstrap/postinstall.sh /bootstrap/postinstall.sh.bkp || echo Done
         - - /bin/bash "${SPACK_ARTIFACTS_ROOT}/postinstall.sh" -fg
-          - spack config --scope site add "packages:all:target:\"target=${SPACK_TARGET_ARCH}\""
+          - spack config --scope site add "packages:all:target:\"[${SPACK_TARGET_ARCH}]\""
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-skylake/spack.yaml]:
```suggestion
          - spack config --scope site add "packages:all:target:[${SPACK_TARGET_ARCH}]"
```

**Review by alalazo — CHANGES_REQUESTED** [review]

**Comment by alalazo** [comment]
Looking at the failure, it seems there is also an issue with the `postinstall.sh` script:
```
/tmp/tmp.LNWh9jz2yU: line 146: /etc/profile.d/spack.sh: No such file or directory
[ ... ]
```

**Review by alalazo — APPROVED** [review]
Fixed the PR!

**Pull request merged**